Panic error with inserting values with commas

#### I receive panic error when I'm trying to write message with commas in fields value.
```
# curl -G 'http://localhost:8086/query' --data-urlencode "q=CREATE DATABASE mmm"
{"results":[{}]}
```
```
# curl -XPOST 'http://localhost:8086/write?db=mmm' --data-binary '@t.g'
curl: (52) Empty reply from server
```
```
# cat t.g
{
  "database" : "mmm",
  "retentionPolicy" : "",
  "points" :  [ {
    "measurement" : "http_status",
    "tags" : {
      "service" : "identity-service",
      "hostname" : "host1",
      "url" : "http://localhost:35774/"
    },
    "timestamp" : "2015-06-16T09:28:26.000Z",
    "fields" : {
      "value_meta" : "{\"error\":\"Pattern match failed! \\\".*v3.0.*\\\" not in \\\"{\\\"version\\\": {\\\"status\\\": \\\"stable\\\", \\\"updated\\\": \\\"2015-03-30T00:00:00Z\\\", \\\"media-types\\\": [{\\\"base\\\": \\\"application/json\\\", \\\"type\\\": \\\"application/json\\\"}], \\\"id\\\": \\\"v3.4\\\", \\\"links\\\": [{\\\"href\\\": \\\"http://localhost:35774/\\\", \\\"rel\\\": \\\"self\\\"}]}}\\\"\"}",
      "value" : 1.0
    }
  } ],
  "tags" : { }
}
```

#### I see stacktrace in log (I added some printf to investigate what happens with these values):
```
field |value|
value |1|
field |value_meta|
value |{"error":"Pattern match failed! ".*v3.0.*" not in "{"version": {"status": "stable\|
field | \"updated\": \"2015-03-30T00:00:00Z\", \"media-types\": [{\"base\": \"application/json\", \"type\": \"application/json\"}], \"id\": \"v3.4\", \"links\": [{\"href\": \"http://localhost:35774/\", \"rel\": \"self\"}]}}\""}"|
panic: unsupported value type during encode fields: <nil>

goroutine 46 [running]:
github.com/influxdb/influxdb/tsdb.(*FieldCodec).EncodeFields(0xc208104180, 0xc20813a0c0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/zielonkam/gocodez/src/github.com/influxdb/influxdb/tsdb/shard.go:582 +0xf3b
github.com/influxdb/influxdb/tsdb.(*Shard).WritePoints(0xc208110240, 0xc2081461b0, 0x1, 0x1, 0x0, 0x0)
	/home/zielonkam/gocodez/src/github.com/influxdb/influxdb/tsdb/shard.go:145 +0x427
github.com/influxdb/influxdb/tsdb.(*Store).WriteToShard(0xc2080376c0, 0x1, 0xc2081461b0, 0x1, 0x1, 0x0, 0x0)
	/home/zielonkam/gocodez/src/github.com/influxdb/influxdb/tsdb/store.go:260 +0xf7
github.com/influxdb/influxdb/cluster.func·002(0x1, 0x1, 0xc2081461b0, 0x1, 0x1)
	/home/zielonkam/gocodez/src/github.com/influxdb/influxdb/cluster/points_writer.go:241 +0xfc
created by github.com/influxdb/influxdb/cluster.(*PointsWriter).writeToShard
	/home/zielonkam/gocodez/src/github.com/influxdb/influxdb/cluster/points_writer.go:271 +0x30b





panic: unsupported value type during encode fields: <nil>

goroutine 43 [running]:
github.com/influxdb/influxdb/tsdb.(*FieldCodec).EncodeFields(0xc2080ed5f0, 0xc2081193b0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/root/.gvm/pkgsets/go1.4.2/global/src/github.com/influxdb/influxdb/tsdb/shard.go:573 +0x8e3
github.com/influxdb/influxdb/tsdb.(*Shard).WritePoints(0xc2080fe240, 0xc2080ed220, 0x1, 0x1, 0x0, 0x0)
	/root/.gvm/pkgsets/go1.4.2/global/src/github.com/influxdb/influxdb/tsdb/shard.go:142 +0x427
github.com/influxdb/influxdb/tsdb.(*Store).WriteToShard(0xc208065800, 0x2, 0xc2080ed220, 0x1, 0x1, 0x0, 0x0)
	/root/.gvm/pkgsets/go1.4.2/global/src/github.com/influxdb/influxdb/tsdb/store.go:256 +0xf7
github.com/influxdb/influxdb/cluster.func·002(0x2, 0x1, 0xc2080ed220, 0x1, 0x1)
	/root/.gvm/pkgsets/go1.4.2/global/src/github.com/influxdb/influxdb/cluster/points_writer.go:213 +0xfc
created by github.com/influxdb/influxdb/cluster.(*PointsWriter).writeToShard
	/root/.gvm/pkgsets/go1.4.2/global/src/github.com/influxdb/influxdb/cluster/points_writer.go:243 +0x2f
```

#### I checked that if I escape commas in field as below then influxdb is working properly:
```
{
  "database" : "mmm",
  "retentionPolicy" : "",
  "points" :  [ {
    "measurement" : "http_status",
    "tags" : {
      "service" : "identity-service",
      "hostname" : "host1",
      "url" : "http://localhost:35774/"
    },
    "timestamp" : "2015-06-16T09:28:26.000Z",
    "fields" : {
      "value_meta" : "{\"error\":\"Pattern match failed! \\\".*v3.0.*\\\" not in \\\"{\\\"version\\\": {\\\"status\\\": \\\"stable\\\"\\, \\\"updated\\\": \\\"2015-03-30T00:00:00Z\\\"\\, \\\"media-types\\\": [{\\\"base\\\": \\\"application/json\\\"\\, \\\"type\\\": \\\"application/json\\\"}]\\, \\\"id\\\": \\\"v3.4\\\"\\, \\\"links\\\": [{\\\"href\\\": \\\"http://localhost:35774/\\\"\\, \\\"rel\\\": \\\"self\\\"}]}}\\\"\"}",
      "value" : 1.0
    }
  } ],
  "tags" : { }
}
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,6 +3,7 @@
 ### Features
 
 - [2650](https://github.com/influxdb/influxdb/pull/2650): Add SHOW GRANTS FOR USER statement. Thanks @n1tr0g.
+- [3013](https://github.com/influxdb/influxdb/issues/3013): Panic error with inserting values with commas
 
 ### Bugfixes
 

--- a/tsdb/points.go
+++ b/tsdb/points.go
@@ -5,6 +5,7 @@ import (
 	"fmt"
 	"hash/fnv"
 	"math"
+	"regexp"
 	"sort"
 	"strconv"
 	"strings"
@@ -54,6 +55,9 @@ type point struct {
 	// binary encoded field data
 	data []byte
 }
+
+// Compile the regex that detects unquoted double quote sequences
+var quoteReplacer = regexp.MustCompile(`([^\\])"`)
 
 var escapeCodes = map[byte][]byte{
 	',': []byte(`\,`),
@@ -604,7 +608,8 @@ func scanFieldValue(buf []byte, i int) (int, []byte) {
 			break
 		}
 
-		if buf[i] == '"' {
+		// If we see a double quote, makes sure it is not escaped
+		if buf[i] == '"' && buf[i-1] != '\\' {
 			i += 1
 			quoted = !quoted
 			continue
@@ -649,6 +654,21 @@ func unescapeString(in string) string {
 		in = strings.Replace(in, esc, b, -1)
 	}
 	return in
+}
+
+// escapeQuoteString returns a copy of in with any double quotes that
+// have not been escaped with escaped quotes
+func escapeQuoteString(in string) string {
+	if strings.IndexAny(in, `"`) == -1 {
+		return in
+	}
+	return quoteReplacer.ReplaceAllString(in, `$1\"`)
+}
+
+// unescapeQuoteString returns a copy of in with any escaped double-quotes
+// with unescaped double quotes
+func unescapeQuoteString(in string) string {
+	return strings.Replace(in, `\"`, `"`, -1)
 }
 
 // NewPoint returns a new point with the given measurement name, tags, fiels and timestamp
@@ -895,7 +915,7 @@ func newFieldsFromBinary(buf []byte) Fields {
 
 		// If the first char is a double-quote, then unmarshal as string
 		if valueBuf[0] == '"' {
-			value = unescapeString(string(valueBuf[1 : len(valueBuf)-1]))
+			value = unescapeQuoteString(string(valueBuf[1 : len(valueBuf)-1]))
 			// Check for numeric characters
 		} else if (valueBuf[0] >= '0' && valueBuf[0] <= '9') || valueBuf[0] == '-' || valueBuf[0] == '.' {
 			value, err = parseNumber(valueBuf)
@@ -955,7 +975,7 @@ func (p Fields) MarshalBinary() []byte {
 			b = append(b, t...)
 		case string:
 			b = append(b, '"')
-			b = append(b, []byte(t)...)
+			b = append(b, []byte(escapeQuoteString(t))...)
 			b = append(b, '"')
 		case nil:
 			// skip

--- a/tsdb/points_test.go
+++ b/tsdb/points_test.go
@@ -453,7 +453,7 @@ func TestParsePointWithStringWithCommas(t *testing.T) {
 			},
 			Fields{
 				"value": 1.0,
-				"str":   "foo,bar", // commas in string value
+				"str":   `foo\,bar`, // commas in string value
 			},
 			time.Unix(1, 0)),
 	)
@@ -469,6 +469,37 @@ func TestParsePointWithStringWithCommas(t *testing.T) {
 			Fields{
 				"value": 1.0,
 				"str":   "foo,bar", // commas in string value
+			},
+			time.Unix(1, 0)),
+	)
+
+}
+
+func TestParsePointEscapedStringsAndCommas(t *testing.T) {
+	// non-escaped comma and quotes
+	test(t, `cpu,host=serverA,region=us-east value="{Hello\"{,}\" World}" 1000000000`,
+		NewPoint(
+			"cpu",
+			Tags{
+				"host":   "serverA",
+				"region": "us-east",
+			},
+			Fields{
+				"value": `{Hello"{,}" World}`,
+			},
+			time.Unix(1, 0)),
+	)
+
+	// escaped comma and quotes
+	test(t, `cpu,host=serverA,region=us-east value="{Hello\"{\,}\" World}" 1000000000`,
+		NewPoint(
+			"cpu",
+			Tags{
+				"host":   "serverA",
+				"region": "us-east",
+			},
+			Fields{
+				"value": `{Hello"{\,}" World}`,
 			},
 			time.Unix(1, 0)),
 	)

